### PR TITLE
Add description and due support to todo lists

### PR DESCRIPTION
**Pull request by bramkragten**
## Proposed change

CleanShot 2023-12-21 at 12 46 29

CleanShot 2023-12-21 at 12 46 46

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/panels/todo/dialog-todo-item-editor.ts [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+    const isCreate = this._params.item === undefined;
+
+    const { dueDate, dueTime } = this._getLocaleStrings(this._due);
+
+    return html`
+      <ha-dialog
+        open
+        @closed=${this.closeDialog}
+        scrimClickAction
+        escapeKeyAction
```
Comment (piitaya) [inline src/panels/todo/dialog-todo-item-editor.ts:96]:
```suggestion
```
Do you think we can allow escape to easily close the popup if the user misclick the checkbox and it opens the dialog?

File: src/panels/lovelace/cards/hui-todo-list-card.ts [hunk 2]
```diff
@@ -209,38 +225,42 @@ export class HuiTodoListCard extends LitElement implements LovelaceCard {
...
-              `
+                  <ha-svg-icon
+                    slot="graphic"
+                    .path=${mdiSort}
+                    .disabled=${unavailable}
+                  >
+                  </ha-svg-icon>
+                </ha-list-item>
+              </ha-button-menu>`
             : nothing}
```
Comment (piitaya) [inline src/panels/lovelace/cards/hui-todo-list-card.ts:255]: Should we hide unchecked header if there is unchecked item?

CleanShot 2023-12-21 at 16 21 43

**Review by piitaya — COMMENTED** [review]
We should check for supported features in the dialog to only show field if the entity support it. Maybe, we should pass the full supported feature instead of `canUpdate` and `canDelete`

File: src/panels/todo/dialog-todo-item-editor.ts (hunk 1 above)
Comment (bramkragten) [reply src/panels/todo/dialog-todo-item-editor.ts:96]: The rule is to not allow it on dialogs with input forms 🤷

File: src/panels/lovelace/cards/hui-todo-list-card.ts (hunk 2 above)
Comment (bramkragten) [reply src/panels/lovelace/cards/hui-todo-list-card.ts:255]: not sure we should hide or just show a message underneath that `There are not items to do!`

File: src/panels/todo/dialog-todo-item-editor.ts [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+              @change=${this._checkedCanged}
+              .disabled=${isCreate || !canUpdate}
+            ></ha-checkbox>
+            <ha-textfield
+              class="summary"
+              name="summary"
+              .label=${this.hass.localize("ui.components.todo.item.summary")}
+              .value=${this._summary}
+              required
+              @change=${this._handleSummaryChanged}
```
Comment (piitaya) [inline src/panels/todo/dialog-todo-item-editor.ts:134]:
```suggestion
              @input=${this._handleSummaryChanged}
```

File: src/panels/todo/dialog-todo-item-editor.ts [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+            TodoListEntityFeature.SET_DESCRIPTION_ON_ITEM
+          )
+            ? html`<ha-textarea
+                class="description"
+                name="description"
+                .label=${this.hass.localize(
+                  "ui.components.todo.item.description"
+                )}
+                .value=${this._description}
+                @change=${this._handleDescriptionChanged}
```
Comment (piitaya) [inline src/panels/todo/dialog-todo-item-editor.ts:152]:
```suggestion
                @input=${this._handleDescriptionChanged}
```

**Review by piitaya — APPROVED** [review]

**Comment by allenporter** [comment]
FYI I've got a pending PR to change the API to support clearing fields https://github.com/home-assistant/core/pull/106208

**Pull request merged**